### PR TITLE
feat(appset): appset git generator verifies GPG signatures

### PR DESCRIPTION
**Pull request by ishitasequeira**
The goal of this PR is for the ApplicationSet Git Generator to verify GPG signatures in the same way that Applications already support this functionality. 

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: reposerver/repository/repository.go [hunk 1]
```diff
@@ -2573,6 +2573,25 @@ func (s *Service) GetGitFiles(_ context.Context, request *apiclient.GitFilesRequ
...
+			log.Errorf("error verifying signature of commit '%s' in repo '%s': %v", revision, repo.Repo, err)
+			return nil, err
+		}
+
+		if cs != "" {
+			vr := gpg.ParseGitCommitVerification(cs)
+			if vr.Result == gpg.VerifyResultUnknown {
+				return nil, fmt.Errorf("UNKNOWN signature: %s", vr.Message)
+			} else {
+				log.Debugf("%s signature from %s key %s", vr.Result, vr.Cipher, gpg.KeyID(vr.KeyID))
```
Comment (pasha-codefresh) [inline reposerver/repository/repository.go:2588]: Here few more cases, like 	
VerifyResultBad     = "Bad"
VerifyResultInvalid = "Invalid", 
I think it also should be an error

File: reposerver/repository/repository.go [hunk 2]
```diff
@@ -2573,6 +2573,25 @@ func (s *Service) GetGitFiles(_ context.Context, request *apiclient.GitFilesRequ
 		return nil, status.Errorf(codes.Internal, "unable to resolve git revision %s: %v", revision, err)
 	}
 
+	if gpg.IsGPGEnabled() && request.VerifyCommit {
```
Comment (pasha-codefresh) [inline reposerver/repository/repository.go:2576]: i would put it in private function, like: verifyCommitSignature, which will just return error

File: applicationset/generators/git.go [hunk 3]
```diff
@@ -58,12 +61,36 @@ func (g *GitGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.Applic
...
+
+	log.Info("===============project=================")
+	log.Info(project)
+
+	appProject := &argoprojiov1alpha1.AppProject{}
+	if err := client.Get(context.TODO(), types.NamespacedName{Name: appSet.Spec.Template.Spec.Project, Namespace: appSet.Namespace}, appProject); err != nil {
+		return nil, fmt.Errorf("error getting project %s: %w", project, err)
+	}
+
+	log.Infof("===============GPG Enabled : %v=================", gpg.IsGPGEnabled())
```
Comment (pasha-codefresh) [inline applicationset/generators/git.go:79]: i would not do such log, can we just replace it to some plain text?

File: applicationset/generators/git.go [hunk 4]
```diff
@@ -58,12 +61,36 @@ func (g *GitGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.Applic
...
+	// we need to verify the signature on the Git revision if GPG is enabled
+	verifyCommit := false
```
Comment (pasha-codefresh) [inline applicationset/generators/git.go:81]: can we just change it to verifyCommit := appProject.Spec.SignatureKeys != nil && len(appProject.Spec.SignatureKeys) > 0 && gpg.IsGPGEnabled()

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

File: applicationset/generators/git.go (hunk 3 above)
Comment (ishitasequeira) [reply applicationset/generators/git.go:79]: I thought I removed this log statements.. I will remove them again.

File: applicationset/generators/git.go (hunk 4 above)
Comment (ishitasequeira) [reply applicationset/generators/git.go:81]: Sure.. Tried keeping it simple and matching to what we have for applications today.

File: reposerver/repository/repository.go (hunk 1 above)
Comment (ishitasequeira) [reply reposerver/repository/repository.go:2588]: I tried keeping the cases similar to what we have today for applications to maintain the same behavior for application-sets

Comment (pasha-codefresh) [reply reposerver/repository/repository.go:2588]: Make sense for me, thanks for explanation

**Review by pasha-codefresh — APPROVED** [review]
LGTM, great job

**Pull request merged**